images/snapshot-release-tags/prune_snapshots.py

```diff
@@ -0,0 +1,85 @@
+#!/usr/bin/env python3
+
+import argparse
+import os
+import re
+import subprocess
+
+import sentry_sdk
+from github import Github
+
+
+sentry_sdk.init(
+    # This cron job only runs once weekly,
+    # so just record all transactions.
+    traces_sample_rate=1.0,
+)
+
+TAG_REF_REGEX = re.compile(r"^refs/tags/(develop-\d{4}-\d{2}-\d{2})$")
+
+
+def main():
+    if "GITHUB_TOKEN" not in os.environ:
+        raise Exception("GITHUB_TOKEN environment is not set")
+
+    parser = argparse.ArgumentParser(
+        prog="prune_snapshots.py",
+        description="Prune expired snapshots",
+    )
+
+    parser.add_argument(
+        "-k",
+        "--keep-last-n",
+        type=int,
+        default=8,
+        help="Prune all but most recent --keep-last-n",
+    )
+    parser.add_argument(
+        "-m",
+        "--mirror-root",
+        default="s3://spack-binaries",
+        help=("Root url of mirror where snapshot binaries are mirrored"),
+    )
+
+    args = parser.parse_args()
+
+    keep_n = args.keep_last_n
+    mirror_root_url = args.mirror_root
+
+    # Use the GitHub API to create a tag for this commit of develop.
+    github_token = os.environ.get("GITHUB_TOKEN")
+    py_github = Github(github_token)
+    py_gh_repo = py_github.get_repo("spack/spack", lazy=True)
+
+    # Get a list of all the tags matching the develop snapshot pattern
+    snapshot_tags = py_gh_repo.get_git_matching_refs("tags/develop-")
+
+    # Sort them so we can prune all but the KEEP_LAST_N most recent
+    pruning_candidates = sorted(snapshot_tags, key=lambda ref: ref.ref)[:-keep_n]
+
+    print("Deleting the following snapshots:")
+    for tag in pruning_candidates:
+        m = TAG_REF_REGEX.search(tag.ref)
+
+        if not m:
+            print(f"Unable to parse {tag.ref}, skipping")
+            continue
+
+        mirror_prefix = m.group(1)
+        url_to_prune = f"{mirror_root_url}/{mirror_prefix}"
+
+        print(f"  Ref: {tag.ref}, Mirror: {url_to_prune}")
+
+        # First, try to delete the mirror associated with the snapshot
+        try:
+            subprocess.run(["aws", "s3", "rm", "--recursive", url_to_prune], check=True)
```

Comment:
An idea for a process for making making sure the cache.spack.io and the mirrors are in sync.

check if mirror has `index.json`
- if yes -> remove `index.json`
- if no -> delete the entire prefix

This will give a buffer time between dropping the tag and deleting the mirror contents. If we run this weekly that should translate to the cache.spack.io page being updated, it uses the the `index.json` to create a global index file.

The timeline for snapshot pruning could be:

-- Prune Cron Runs @ 2023/01/01 0100 UTC
- Delete Tag `develop-XXXX`
- Delete `develop-XXXX/build_cache/index.json`

-- Generate cache.spack.io @ 2023/01/02 0100 UTC
- Create global index
- push new website without mirrors containing no `index.json`

-- Prune Cron Runs @ 2023/01/08 0100 UTC
- Delete prefix `develop-XXXX/`